// stryker-mocha-runner/src/StrykerMochaReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_runner_1 = require("stryker-api/test_runner");
var logging_1 = require("stryker-api/logging");
var Timer_1 = require("./Timer");
var StrykerMochaReporter = /** @class */ (function () {
    function StrykerMochaReporter(runner) {
        this.runner = runner;
        this.log = logging_1.getLogger(StrykerMochaReporter.name);
        this.timer = new Timer_1.default();
        this.passedCount = 0;
        this.registerEvents();
        StrykerMochaReporter.CurrentInstance = this;
    }
    StrykerMochaReporter.prototype.registerEvents = function () {
        var _this = this;
        this.runner.on('start', function () {
            _this.passedCount = 0;
            _this.timer.reset();
            _this.runResult = {
                status: test_runner_1.RunStatus.Error,
                tests: [],
                errorMessages: []
            };
            _this.log.debug('Starting Mocha test run');
        });
        this.runner.on('pass', function (test) {
            _this.runResult.tests.push({
                status: test_runner_1.TestStatus.Success,
                name: test.fullTitle(),
                timeSpentMs: _this.timer.elapsedMs()
            });
            _this.passedCount++;
            _this.timer.reset();
        });
        this.runner.on('fail', function (test, err) {
            _this.runResult.tests.push({
                status: test_runner_1.TestStatus.Failed,
                name: test.fullTitle(),
                timeSpentMs: _this.timer.elapsedMs(),
                failureMessages: [err.message]
            });
            if (!_this.runResult.errorMessages) {
                _this.runResult.errorMessages = [];
            }
            _this.runResult.errorMessages.push(err.message);
            if (_this.log.isTraceEnabled()) {
                _this.log.trace("Test failed: " + test.fullTitle() + ". Error: " + err.message);
            }
        });
        this.runner.on('end', function () {
            _this.runResult.status = test_runner_1.RunStatus.Complete;
            _this.log.debug('Mocha test run completed: %s/%s passed', _this.passedCount, _this.runResult.tests.length);
        });
    };
    return StrykerMochaReporter;
}());
exports.default = StrykerMochaReporter;
