// stryker/src/ReporterOrchestrator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var logging_1 = require("stryker-api/logging");
var report_1 = require("stryker-api/report");
var BroadcastReporter_1 = require("./reporters/BroadcastReporter");
var ClearTextReporter_1 = require("./reporters/ClearTextReporter");
var DashboardReporter_1 = require("./reporters/DashboardReporter");
var DotsReporter_1 = require("./reporters/DotsReporter");
var EventRecorderReporter_1 = require("./reporters/EventRecorderReporter");
var ProgressAppendOnlyReporter_1 = require("./reporters/ProgressAppendOnlyReporter");
var ProgressReporter_1 = require("./reporters/ProgressReporter");
function registerDefaultReporters() {
    report_1.ReporterFactory.instance().register('progress-append-only', ProgressAppendOnlyReporter_1.default);
    report_1.ReporterFactory.instance().register('progress', ProgressReporter_1.default);
    report_1.ReporterFactory.instance().register('dots', DotsReporter_1.default);
    report_1.ReporterFactory.instance().register('clear-text', ClearTextReporter_1.default);
    report_1.ReporterFactory.instance().register('event-recorder', EventRecorderReporter_1.default);
    report_1.ReporterFactory.instance().register('dashboard', DashboardReporter_1.default);
}
registerDefaultReporters();
var ReporterOrchestrator = /** @class */ (function () {
    function ReporterOrchestrator(options) {
        this.options = options;
        this.log = logging_1.getLogger(ReporterOrchestrator.name);
    }
    ReporterOrchestrator.prototype.createBroadcastReporter = function () {
        var _this = this;
        var reporters = [];
        var reporterOption = this.options.reporters;
        if (reporterOption) {
            reporterOption.forEach(function (reporterName) { return reporters.push(_this.createReporter(reporterName)); });
        }
        else {
            this.log.warn("No reporter configured. Please configure one or more reporters in the (for example: reporters: ['progress'])");
            this.logPossibleReporters();
        }
        return new BroadcastReporter_1.default(reporters);
    };
    ReporterOrchestrator.prototype.createReporter = function (name) {
        if (name === 'progress' && !process.stdout.isTTY) {
            this.log.info('Detected that current console does not support the "progress" reporter, downgrading to "progress-append-only" reporter');
            return { name: 'progress-append-only', reporter: report_1.ReporterFactory.instance().create('progress-append-only', this.options) };
        }
        else {
            return { name: name, reporter: report_1.ReporterFactory.instance().create(name, this.options) };
        }
    };
    ReporterOrchestrator.prototype.logPossibleReporters = function () {
        var possibleReportersCsv = '';
        report_1.ReporterFactory.instance().knownNames().forEach(function (name) {
            if (possibleReportersCsv.length) {
                possibleReportersCsv += ', ';
            }
            possibleReportersCsv += name;
        });
        this.log.warn("Possible reporters: " + possibleReportersCsv);
    };
    return ReporterOrchestrator;
}());
exports.default = ReporterOrchestrator;
